libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = std::set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& _seen,
+	TypeSet& _oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen.count(id))
+			continue;
+		_seen.insert(id);
+		bool includeType = _type->category() != Type::Category::Struct;
+		storageOversizedSubtypesInner(subtype, includeType, _seen, _oversizedSubtypes);
+		_seen.erase(id);
+	}
+
+	if (_includeType && _type->storageSizeUpperBound() >= bigint(1) << 64)
+		_oversizedSubtypes.insert(_type);
+}
+
+}
+
+vector<Type const*> Type::storageOversizedSubtypes() const
+{
+	set<string> seen;
+	TypeSet oversized(compareTypes);
+	storageOversizedSubtypesInner(this, false, seen, oversized);
+	vector<Type const*> res;
 	copy(oversized.cbegin(), oversized.cend(), back_inserter(res));
 	return res;
 }
 
+BoolResult Type::validForLocation(DataLocation _loc, set<string>* _seen) const
+{
+	if (_seen == nullptr)
+	{
+		set<string> seen;
+		return validForLocation(_loc, &seen);
+	}
+
+	for (Type const* subtype: subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen->count(id))
+			continue;
+		_seen->insert(id);
+		BoolResult result = subtype->validForLocation(_loc, _seen);
+		_seen->erase(id);
+		if (!result)
+			return result;
```

Comment:
This would read much better with a name that describes the condition:
```suggestion
		BoolResult valid = subtype->validForLocation(_loc, _seen);
		_seen->erase(id);
		if (!valid)
			return valid;
```

Reply:
OK, although
```
if (!valid)
    return valid;
```
might be misread too as returning something positive.

Reply:
Yeah, it's less of a problem with plain `bool`s because I'd normally just do `return false`. But here you have to pass on the message. Calling `BoolResult::err()` here and giving it the message would make it clearer.

I still think it's a better name than `result` which is usually adds no more information than just calling the variable `x` :)

Reply:
> I still think it's a better name than `result` which is usually adds no more information than just calling the variable `x` :)

I would say, just for the sake of argument, that `result` or `res` might may sense for a variable which keeps the function return value (think of `vector f() { vector<int> res; .... ; return res; }`). But in the given context it is not an appropriate name, sure.